pkg/verify/verify.go

```diff
@@ -0,0 +1,469 @@
+package verify
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/client-go/transport"
+
+	"github.com/golang/glog"
+	"golang.org/x/crypto/openpgp"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"github.com/openshift/cluster-version-operator/pkg/payload"
+)
+
+// Interface performs verification of the provided content. The default implementation
+// in this package uses the container signature format defined at https://github.com/containers/image
+// to authenticate that a given release image digest has been signed by a trusted party.
+type Interface interface {
+	Verify(ctx context.Context, releaseDigest string) error
+}
+
+type rejectVerifier struct{}
+
+func (rejectVerifier) Verify(ctx context.Context, releaseDigest string) error {
+	return fmt.Errorf("verification is not possible")
+}
+
+// Reject fails always fails verification.
+var Reject Interface = rejectVerifier{}
+
+// LoadFromPayload looks for a config map in the v1 API group within within the provided update
+// with the annotation "release.openshift.io/verification-config-map". Only the first payload
+// entry with the annotation is used.
+//
+// The presence of this config map instructs the CVO to verify updates before they are downloaded.
+//
+// The keys within the config map define how verification is performed:
+//
+// verifier-public-key-*: One or more GPG public keys in ASCII form that must have signed the
+//                        release image by digest.
+//
+// store-*: A URL (scheme file://, http://, or https://) location that contains signatures. These
+//          signatures are in the atomic container signature format. The URL will have the digest
+//          of the image appended to it as "<STORE>/<ALGO>=<DIGEST>/signature-<NUMBER>" as described
+//          in the container image signing format. The docker-image-manifest section of the
+//          signature must match the release image digest. Signatures are searched starting at
+//          NUMBER 1 and incrementing if the signature exists but is not valid. The signature is a
+//          GPG signed and encrypted JSON message.
+//
+// See https://github.com/containers/image/blob/ab49b0a48428c623a8f03b41b9083d48966b34a9/docs/signature-protocols.md
+// for a description of the signature store
+//
+// The returned verifier will require that any new release image will only be considered verified
+// if each provided public key has signed the release image digest. The signature may be in any
+// store and the lookup order is internally defined.
+//
+func LoadFromPayload(update *payload.Update) (Interface, error) {
```

Comment:
nit it would be better if we did not the errors, and use the `github.com/pkg/errors` for wrapping the errors.